sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -0,0 +1,206 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.orc
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.io._
+import org.apache.orc.mapred.{OrcList, OrcMap, OrcStruct, OrcTimestamp}
+import org.apache.orc.storage.serde2.io.{DateWritable, HiveDecimalWritable}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecificInternalRow
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+private[orc] class OrcDeserializer(
+    dataSchema: StructType,
+    requiredSchema: StructType,
+    maybeMissingSchema: Option[StructType]) {
+
+  private[this] val mutableRow = new SpecificInternalRow(requiredSchema.map(_.dataType))
+
+  private[this] val unwrappers = requiredSchema.fields.map(f => unwrapperFor(f.dataType))
+
+  def deserialize(writable: OrcStruct): InternalRow = {
+    convertOrcStructToInternalRow(writable, dataSchema, requiredSchema,
+      maybeMissingSchema, Some(unwrappers), Some(mutableRow))
+  }
+
+  /**
+   * Convert Apache ORC OrcStruct to Apache Spark InternalRow.
+   * If internalRow is not None, fill into it. Otherwise, create a SpecificInternalRow and use it.
+   */
+  private[this] def convertOrcStructToInternalRow(
+      orcStruct: OrcStruct,
+      dataSchema: StructType,
+      requiredSchema: StructType,
+      missingSchema: Option[StructType] = None,
+      valueUnwrappers: Option[Seq[(Any, InternalRow, Int) => Unit]] = None,
+      internalRow: Option[InternalRow] = None): InternalRow = {
+    val mutableRow = internalRow.getOrElse(new SpecificInternalRow(requiredSchema.map(_.dataType)))
+    val unwrappers =
+      valueUnwrappers.getOrElse(requiredSchema.fields.map(_.dataType).map(unwrapperFor).toSeq)
+    var i = 0
+    val len = requiredSchema.length
+    val names = orcStruct.getSchema.getFieldNames
+    while (i < len) {
+      val name = requiredSchema(i).name
+      val writable = if (missingSchema.isEmpty || missingSchema.get.getFieldIndex(name).isEmpty) {
```

Comment:
I think this check is only needed for top-level columns.

Reply:
Can we strictly follow the style in `OrcFileFormat.unwrapOrcStructs`? i.e. no method like `convertOrcStructToInternalRow`, top-level columns and struct fields are handled with different while loops.